test/integration/test_config_defaults.py

```diff
@@ -0,0 +1,1107 @@
+      in some form at load time; otherwise it should not be listed as a default).
+
+    Configuration options NOT tested:
+    - config_dir (value overridden for testing)
+    - data_dir (value overridden for testing)
+    - new_file_path (value overridden for testing)
+    - logging (mapping loaded in config/; TODO)
+    - dependency_resolution (nested properties; TODO)
+    - job_config (no obvious testable defaults)
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(ConfigDefaultsTestCase, self).__init__(*args, **kwargs)
+        # use lazy loading for attributes below
+        self._root_dir = None
+        self._config_dir = None
+        self._data_dir = None
+        self._tool_data_path = None
+
+    def get_default(self, key):
+        # Returns default value or None (because if default value is not
+        # specified in the schema, we expect None  by default)
+        return self._app.config.appschema[key].get('default', None)
+
+    def get_csv_default_as_list(self, key):
+        # Use when schema default is a comma-separated values list
+        # which is converted to a list
+        return self._app.config.appschema[key]['default'].split(',')
+
+    def get_default_in_root_dir(self, key):
+        # Use when schema default is resolved with respect to the installation's root directory
+        self._root_dir = self._root_dir or self._app.config.root
+        return self._resolve(self._root_dir, self.get_default(key))
+
+    def get_default_in_config_dir(self, key):
+        # Use when schema default is resolved with respect to the value of 'config_dir'
+        self._config_dir = self._config_dir or self._app.config.config_dir
+        return self._resolve(self._config_dir, self.get_default(key))
+
+    def get_default_in_data_dir(self, key):
+        # Use when schema default is resolved with respect to the value of 'data_dir'
+        self._data_dir = self._data_dir or self._app.config.data_dir
+        return self._resolve(self._data_dir, self.get_default(key))
+
+    def get_default_in_tool_data_path(self, key):
+        # Use when schema default is resolved with respect to the value of 'tool_data_path'
+        self._tool_data_path = self._tool_data_path or self._app.config.tool_data_path
+        return self._resolve(self._tool_data_path, self.get_default(key))
+
+    def _resolve(self, parent, child):
+        return os.path.join(parent, child) if child else parent
+
+    # def test_default_database_connection(self):
+    # TODO: untestable; refactor config/__init__ to test
+
+    # def test_default_database_engine_option_pool_size(self):
+    # TODO: default value overridden for tests runnign on non-sqlite databases
+
+    # def test_default_database_engine_option_max_overflow(self):
+    # TODO: default value overridden for tests runnign on non-sqlite databases
+
+    def test_default_database_engine_option_pool_recycle(self):
```

Comment:
Could you parametrize all the `test_.*` methods using @pytest.mark.parametrize ?
I'd build a list of config strings up front (or even discover them from the schema) and then parametrize a generic function, i.e
```python
CONFIG_PROPERTIES = [
    'database_engine_option_server_side_cursors',
    'database_engine_option_pool_recycle',
]


@pytest.mark.parametrize('config_property', CONFIG_PROPERTIES)
def test_default_config_property(config_property):
    expect = self.get_default(config_property)
    assert expect == getattr(self._app.config, config_property)
```

If you can auto-discover these from the schema: even better :)

Reply:
Probably yes. Marking this as WIP for now, while I'm consolidating tests.

Reply:
@mvdbeek how's this? I can barely stop myself from adding 'one test to find them' to the commit message :)